src/index.ts

```diff
@@ -1,65 +1,238 @@
+    (req.body as Buffer)?.toString('utf8') ?? '<no request body>';
+  logger.debug(
+    {
+      method: req.method,
+      url: req.url,
+      request: bodyString,
+    },
+    `Request to ${req.method} ${req.url}: ${bodyString}`
   );
+  return req;
+}
 
-  server.addContentTypeParser('application/json', (req, payload, done) => {
-    done(null, payload);
+function readPassthrough(readable: Readable, onBody: (body: Buffer) => void) {
+  const chunks: Buffer[] = [];
+  const transform = new Transform({
+    autoDestroy: true,
+    emitClose: true,
+    transform(chunk, _encoding, callback) {
+      chunks.push(chunk);
+      callback(null, chunk);
+    },
+  });
+  transform.on('close', () => {
+    const body = chunks.length === 1 ? chunks[0] : Buffer.concat(chunks);
+    onBody(body);
   });
+  return readable.pipe(transform);
+}
+
+function handleProxyResponse(
+  request: FastifyRequest,
+  reply: FastifyReply,
+  response: Readable
+) {
+  // Handle request/response logging
+  if (ENV.LOG_REQUESTS) {
+    request = addRequestLogging(request);
+  }
+  if (ENV.LOG_RESPONSES) {
+    response = addResponseLogging(request, reply, response);
+  }
+
+  // Handle tx-multicast config
+  if (
+    isTxMulticastEnabled() &&
+    request.method === 'POST' &&
+    request.url === '/v2/transactions'
+  ) {
+    postTxMulticast(request, reply, response).catch((error) => {
+      logger.error(error, `Error performing tx-multicast: ${error}`);
+    });
+  }
+
+  // Handle custom cache-control header config
+  const cacheControl = getCacheControlHeader(reply.statusCode, request.url);
+  if (cacheControl) {
+    reply.header('Cache-Control', cacheControl);
+  }
+
+  reply.send(response);
+}
 
-  server.addContentTypeParser('*', (req, payload, done) => {
-    done(null, payload);
+export async function startServer(): Promise<{
+  address: string;
+  server: FastifyInstance;
+  promServer: FastifyInstance;
+}> {
+  const server = fastify({
+    logger,
+    bodyLimit: ENV.MAX_REQUEST_BODY_SIZE,
+    forceCloseConnections: true,
   });
 
-  await server.register(proxy, {
-    upstream: `http://${ENV.STACKS_CORE_PROXY_HOST}:${ENV.STACKS_CORE_PROXY_PORT}/v2/transactions`,
-    prefix: '/v2/transactions',
-    proxyPayloads: false,
-    preHandler: async (request, reply) => {
-      const body = request.body;
-    },
-    replyOptions: {
-      onResponse(request, reply, res) {
-        reply.send(res);
-      },
-    },
+  // Setup Prometheus metrics, set endpoint to null because we're exposing it via a separate server
+  await server.register(FastifyMetrics, { endpoint: null });
+
+  await server.register(cors, {
+    preflightContinue: false, // Do _not_ pass the CORS preflight OPTIONS request to stacks-node because its http server doesn't properly support it
+    strictPreflight: false, // Unnecessarily strict CORS preflight checks
+    methods: ALLOW_REQUEST_METHODS,
+    allowedHeaders: ALLOW_REQUEST_HEADERS,
+    exposedHeaders: ALLOW_RESPONSE_HEADERS,
```

Comment:
Forewarning, I uncovered a lot of information writing up this comment. Some of it I wish I didn't discover haha

# Original comment
I think we can remove the headers being set here, since it's configurable at the infrastructure layer.

# BUT...
After doing some digging, I found that the Stacks Blockchain code actually _does_ set CORS-related headers for all requests (not just pre-flight requests. [See here for context.](https://github.com/stacks-network/stacks-blockchain/blob/02dabb88b858b6bf028efadc179e1ef545dfad86/src/net/http.rs#L1172-L1176) Previously I thought CORS requests to the HTTP server weren't working due to a lack of support or something similar, but instead it seems it's because the HTTP server is already setting these headers.

This makes me think that maybe CORS headers set at the app-level would override CORS headers set at the infra level. It would make sense, since back when we first tried routing `/v2` traffic directly to the stacks-node, the CORS headers all reflected the aforementioned linked code, and not the CORS setting I had configured in the infrastructure.

Or rather, the CORS settings I had _thought_ were configured in the infrastructure...I decided to do some experiments in the dev cluster.

# The First Discovery...
While experimenting with different CORS settings at the infrastructure level for `/v2` traffic being sent directly to a stacks blockchain node, I was able to get the settings I configured there to return on OPTIONS requests, invalidating my previous thought of app-level CORS settings overriding infra-level CORS settings.
So why were app-level CORS settings taking precedence when we previously tried routing `/v2` traffic directly to the stacks-node back in March, but the infra-level CORS settings are taking precedence in my experiment?

**It's because Istio does not set defaults for CORS settings**; they all must be explicitly laid out. And in this experiment I set explicit CORS settings in Istio.
Some context -- When configuring CORS for other services, like the stacks-blockchain-api, TM-API, etc, I would only ever need to configure the allowed headers in Istio. The allowed methods would always return `GET,HEAD,PUT,PATCH,POST,DELETE` and the allowed origin would return the address of the origin site sending the CORS preflight request, so everything would work out.
This gave the false sense that Istio was setting implicit CORS defaults, and their documentation on this is obscure enough to support this notion. So when I explicitly configure additional CORS settings at the infra-level which I had not needed to do before, it was unexpected to actually see them come through and work for requests going direct to a stacks blockchain node.

# The Second Discovery...
So if Istio is not setting default CORS values, then why do the API services return `GET,HEAD,PUT,PATCH,POST,DELETE` for allowed methods and an allowed origin matching the one making the CORS request?
After my last bout of digging, I found it's because the CORS packages the [stacks-blockchain-api](https://github.com/expressjs/cors/blob/master/lib/index.js#L10) and [tm-api](https://github.com/fastify/fastify-cors/blob/master/index.js#L11) uses set the allowed methods and allowed origin by default.

So in summary:

* The CORS settings in use for all of our services is a hodgepodge mixture of app-level and infra-level settings
* To only use CORS settings from the infra-level they all need to be explicitly defined at the infra level, as Istio does not set defaults.

This invariably means CORS can be fixed for the stacks-blockchain without a code change and without another app in front of it proxying requests.

To confirm this, I have `/v2` routing directly to stacks blockchain nodes in the dev cluster right now, with all CORS settings configured at the infra-level, and it seems the Explorer is working as expected (@timstackblock can you verify with the Explorer pointing to `https://api.dev.hiro.so` ?).
If I revert to the CORS settings we previouisly had configured in Istio (only allowed headers defined), then address pages break in the explorer.

Assuming the tests go well, this discovery seems to weaken the need for a proxy service like this, at least for Hiro's purposes. I have an idea for how we can propagate `/v2` POSTs to API writers as well as the infra-level, but it will need to be tested.

Reply:
Mostly good news! No worries, this project wasn't _too_ much time. And it's always available for us to use later if needed for whatever reason. Thanks for the researching and debugging.